samza-core/src/main/java/org/apache/samza/execution/PlanJsonGenerator.java

```diff
@@ -0,0 +1,266 @@
+ */
+
+package org.apache.samza.execution;
+
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.OutputStream;
+import org.apache.samza.operators.StreamGraphImpl;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.PartialJoinOperatorSpec;
+import org.apache.samza.operators.spec.SinkOperatorSpec;
+import org.codehaus.jackson.map.ObjectMapper;
+
+/**
+ * This class generates the json representation of the plan.
+ */
+public class PlanJsonGenerator {
+
+  private static abstract class GraphNode {
+    private Integer nodeId;
+    private Set<Integer> successors = new HashSet<>();
+
+    public void setSuccessors(Set<Integer> successors) {
+      this.successors = successors;
+    }
+
+    public void setNodeId(Integer nodeId) {
+      this.nodeId = nodeId;
+    }
+
+    public Integer getNodeId() {
+      return nodeId;
+    }
+
+    public Set<Integer> getSuccessors() {
+      return successors;
+    }
+  }
+
+  public static final class OpNode extends GraphNode {
+    private String jobName;
+    private String jobId;
+    private String opName;
+    private Set<Integer> opIds = new HashSet<>();
+
+    public String getJobName() {
+      return jobName;
+    }
+
+    public void setJobName(String jobName) {
+      this.jobName = jobName;
+    }
+
+    public String getJobId() {
+      return jobId;
+    }
+
+    public void setJobId(String jobId) {
+      this.jobId = jobId;
+    }
+
+    public String getOpName() {
+      return opName;
+    }
+
+    public void setOpName(String opName) {
+      this.opName = opName;
+    }
+
+    public Set<Integer> getOpIds() {
+      return opIds;
+    }
+
+    public void setOpIds(Set<Integer> opIds) {
+      this.opIds = opIds;
+    }
+  }
+
+  public static final class StreamNode extends GraphNode {
+    private String streamId;
+    private String system;
+    private String physicalName;
+    private int partition;
+
+    public String getStreamId() {
+      return streamId;
+    }
+
+    public void setStreamId(String streamId) {
+      this.streamId = streamId;
+    }
+
+    public String getSystem() {
+      return system;
+    }
+
+    public void setSystem(String system) {
+      this.system = system;
+    }
+
+    public String getPhysicalName() {
+      return physicalName;
+    }
+
+    public void setPhysicalName(String stream) {
+      this.physicalName = stream;
+    }
+
+    public int getPartition() {
+      return partition;
+    }
+
+    public void setPartition(int partition) {
+      this.partition = partition;
+    }
+  }
+
+  public static final class GraphNodes {
```

Comment:
It's a little strange that there are public classes for all these nodes that are only there to generate the Plan representation.